packages/terra-disclosure-manager/src/DisclosureManager.jsx

```diff
@@ -0,0 +1,401 @@
+      props: data.content.props,
+      component: data.content.component,
+    };
+
+    this.setState(newState);
+  }
+
+  popDisclosure() {
+    const newState = DisclosureManager.cloneDisclosureState(this.state);
+
+    newState.disclosureComponentData[newState.disclosureComponentKeys.pop()] = undefined;
+
+    this.setState(newState);
+  }
+
+  closeDisclosure() {
+    this.setState({
+      disclosureIsOpen: false,
+      disclosureIsFocused: false,
+      disclosureIsMaximized: false,
+      disclosureSize: undefined,
+      disclosureComponentKeys: [],
+      disclosureComponentData: {},
+    });
+  }
+
+  requestDisclosureFocus() {
+    this.setState({
+      disclosureIsFocused: true,
+    });
+  }
+
+  releaseDisclosureFocus() {
+    this.setState({
+      disclosureIsFocused: false,
+    });
+  }
+
+  maximizeDisclosure() {
+    this.setState({
+      disclosureIsMaximized: true,
+    });
+  }
+
+  minimizeDisclosure() {
+    this.setState({
+      disclosureIsMaximized: false,
+    });
+  }
+
+  resolveDismissPromise(key) {
+    const resolve = this.onDismissResolvers[key];
+    if (resolve) {
+      resolve();
+    }
+    this.onDismissResolvers[key] = undefined;
+  }
+
+  /**
+   * This function resolves the dismiss checks in sequence. The Promise will either resolve if all checks resolve or
+   * reject on the check rejection. This ensures that checks do not occur for components after the first rejection.
+   */
+  resolveDismissChecksInSequence(keys) {
+    return new Promise((resolve, reject) => {
+      if (!keys.length) {
+        resolve();
+        return;
+      }
+
+      const key = keys.pop();
+
+      this.generatePopFunction(key)()
+        .then(() => {
+          this.resolveDismissChecksInSequence(keys).then(resolve).catch(reject);
+        })
+        .catch(() => {
+          reject();
+        });
+    });
+  }
+
+  safelyCloseDisclosure() {
+    const disclosureKeys = Object.assign([], this.state.disclosureComponentKeys);
+
+    /**
+     * Before closing the disclosure, the dismiss checks for components in the stack are
+     * executed in stack order. Components will be dismissed in order up until a rejection occurs, at which point
+     * the blocking component will be presented.
+     */
+    return this.resolveDismissChecksInSequence(disclosureKeys).then(() => {
+      this.dismissChecks = {};
+      this.closeDisclosure();
+    });
+  }
+
+  generatePopFunction(key) {
+    return () => {
+      let promiseRoot = Promise.resolve();
+
+      const dismissCheck = this.dismissChecks[key];
+      if (dismissCheck) {
+        promiseRoot = dismissCheck();
+      }
+
+      return promiseRoot
+        .then(() => {
+          this.dismissChecks[key] = undefined;
+          this.resolveDismissPromise(key);
+        })
+        .then(() => {
+          this.popDisclosure();
+        });
+    };
+  }
+
+  renderContentComponents() {
+    const { children, app } = this.props;
+
+    const appDelegate = {};
```

Comment:
Should this clone app?

Reply:
Whoops, nice catch

Reply:
Same question in renderDisclosureComponents()

Reply:
We could, but we're providing new functions for all AppDelegate APIs so nothing would actually be carried forward.